### PR TITLE
Add submodule `core/`

### PR DESCRIPTION
**Pull request by WeilerP**
## New

Adds the submodule `core/` intended to host code applicable to multiple parts of the code base. The following files are added:

* `_anndata.py` is intended to host code related to manipulating AnnData objects. As of now, the functions `get_modality`, `make_dense`, `make_sparse` and `set_modality` are implemented.
  * `get_modality` allows easily retrieving any modality from `X`, `layers` as well as `obsm`.
  * `make_dense` densifies a specified sparse AnnData entries.
  * `make_sparse` makes specified AnnData entries sparse.
  * `set_modality` allows setting a modality of an AnnData object to a new value.
* `_arithmetic.py` hosts code to perform arithemtic operations. As of now, the functions `clipped_log`, `invert`, `prod_sum` and `sum` are implemented.
  * `clipped_log` logarithmizes an array with values clipped to a specified interval.
  * `invert` allows inverting an array with entries equal to zero.
  * `prod_sum` implements muliplying two arrays and summing the result along a given axis.
  * `sum` sums an array (dense or sparse) over a given axis.
* `_linear_models.py` hosts code to fit linear models. As of now, the class `LinearRegression` is implemented. It largely follows the implementation of `leastsq_NxN` in `scvelo/tools/optimization.py`.
  * `LinearRegression` allows fitting a linear model per feature on extreme quantiles.
* `_metrics.py` is intended to host implementations of different metrics. As of now, the l2 norm has been implemented.
  * `l2_norm` takes the l2 norm of a given array (dense or sparse) along a given axis.
* `_models.py` contains implementations of dynamical systems. As of now, the splicing model has been implemented.
* The functions are unit tested in `scvelo/core/tests`

## Changes

* `sum_obs` and `sum_var` are combined into `sum`.
* `prod_sum_obs`, `prod_sum_var` are combined into `prod_sum`.
* `scvelo/tools/utils.py::norm`, `scvelo/tools/utils.py::vector_norm` are combined into `scvelo/core/_metrics::l2_norm`
* Pytest checks files of the format `test_*.py`

### CODE REVIEW

File: scvelo/core/_models.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        c = (self.alpha - self.u0 * self.beta) * invert(self.gamma - self.beta)
+        spliced = (
+            self.s0 * exps + self.alpha / self.gamma * (1 - exps) + c * (exps - expu)
+        )
+
+        if with_keys:
+            return {"u": unspliced, "s": spliced}
+        else:
+            if isinstance(t, np.ndarray) and t.ndim == 2:
+                return np.stack([unspliced, spliced], axis=2)
```
Comment (VolkerBergen) [inline scvelo/core/_models.py:97]: why not just `return unspliced, spliced`? Do we need stacking, to then later unstack?

Comment (WeilerP) [reply scvelo/core/_models.py:97]: We need it stacked to generalise the recovery of the dynamics and make it (to a large extend) model-independent.

File: scvelo/datasets.py [hunk 2]
```diff
@@ -177,7 +179,18 @@ def draw_poisson(n):
         return np.insert(t, 0, 0)  # prepend t0=0
 
     def simulate_dynamics(tau, alpha, beta, gamma, u0, s0, noise_model, noise_level):
-        ut, st = mRNA(tau, u0, s0, alpha, beta, gamma)
+        state = SplicingDynamics(
+            alpha=alpha, beta=beta, gamma=gamma, initial_state=[u0, s0]
+        ).get_solution(tau)
```
Comment (VolkerBergen) [inline scvelo/datasets.py:184]: via unstacked?

File: scvelo/plotting/simulation.py [hunk 3]
```diff
@@ -18,8 +20,18 @@ def get_dynamics(adata, key="fit", extrapolate=False, sorted=False, t=None):
         t = adata.obs[f"{key}_t"].values if key == "true" else adata.layers[f"{key}_t"]
 
     tau, alpha, u0, s0 = vectorize(np.sort(t) if sorted else t, t_, alpha, beta, gamma)
-    ut, st = mRNA(tau, u0, s0, alpha, beta, gamma)
-
+    state = SplicingDynamics(
+        alpha=alpha, beta=beta, gamma=gamma, initial_state=[u0, s0]
+    ).get_solution(tau)
```
Comment (VolkerBergen) [inline scvelo/plotting/simulation.py:25]: via unstacked?

File: scvelo/plotting/simulation.py [hunk 4]
```diff
@@ -51,7 +63,18 @@ def compute_dynamics(
 
     tau, alpha, u0, s0 = vectorize(np.sort(t) if sort else t, t_, alpha, beta, gamma)
 
-    ut, st = mRNA(tau, u0, s0, alpha, beta, gamma)
+    state = SplicingDynamics(
+        alpha=alpha, beta=beta, gamma=gamma, initial_state=[u0, s0]
+    ).get_solution(tau)
```
Comment (VolkerBergen) [inline scvelo/plotting/simulation.py:68]: via unstacked?

Comment (WeilerP) [reply scvelo/plotting/simulation.py:68]: dito

File: scvelo/datasets.py (hunk 2 above)
Comment (WeilerP) [reply scvelo/datasets.py:184]: Yes!

File: scvelo/plotting/simulation.py (hunk 3 above)
Comment (WeilerP) [reply scvelo/plotting/simulation.py:25]: dito

**Comment by WeilerP** [comment]
I'll close this PR and open a new one s.t. the commit history isn't so cluttered.

**Pull request closed without merging**